hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/HoodieIndexUtils.java

```diff
@@ -311,15 +313,74 @@ private static <R> Option<HoodieRecord<R>> mergeIncomingWithExistingRecordWithEx
       return Option.of(result);
     }
 
+    // At this point, result.getData() contains a partial record update.
+    IndexedRecord existingRecord = existing.toIndexedRecord(existingSchema, config.getProps())
```

Comment:
Let's hold off any changes to the non-conventional merging logic that goes through merger or does not go through the file group reader.  We can simplify the logic along with unifying the reader path.